### PR TITLE
Adding methods of A?_mul_B? when A and B are both diagonal

### PR DESCRIPTION
**Pull request by YingboMa**
This will fix the issue of multiplying two diagonal matrices but the result is a full matrix.

Before
```julia
julia> D
3×3 Diagonal{Complex{Float64}}:
 0.236033+0.346517im           ⋅                    ⋅          
          ⋅           0.251662+0.986666im           ⋅          
          ⋅                    ⋅           0.251379+0.0203749im

julia> D * D'
3×3 Array{Complex{Float64},2}:
 0.175786+0.0im      0.0+0.0im        0.0+0.0im
      0.0+0.0im  1.03684+0.0im        0.0+0.0im
      0.0+0.0im      0.0+0.0im  0.0636066+0.0im
```

After
```julia
julia> D                               
3×3 Diagonal{Complex{Float64}}:        
 0.236033+0.346517im           ⋅                    ⋅                          
          ⋅           0.251662+0.986666im           ⋅                          
          ⋅                    ⋅           0.251379+0.0203749im                

julia> D * D'                          
3×3 Diagonal{Complex{Float64}}:        
 0.175786+0.0im          ⋅                ⋅                                    
          ⋅      1.03684+0.0im            ⋅                                    
          ⋅              ⋅      0.0636066+0.0im 
```

### CODE REVIEW

**Comment by jiahao** [comment]
Could use some tests

File: base/linalg/diagonal.jl [hunk 1]
```diff
@@ -185,20 +185,23 @@ function A_mul_B!(D::Diagonal, B::UnitUpperTriangular)
     UpperTriangular(B.data)
 end
 
+Ac_mul_B(D::Diagonal, B::Diagonal) = Diagonal(conj(D.diag) .* B.diag)
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: Perhaps `conj.(...)` should replace `conj(...)`?

File: base/linalg/diagonal.jl [hunk 2]
```diff
@@ -185,20 +185,23 @@ function A_mul_B!(D::Diagonal, B::UnitUpperTriangular)
...
 Ac_mul_B(A::AbstractTriangular, D::Diagonal) = A_mul_B!(ctranspose(A), D)
 function Ac_mul_B(A::AbstractMatrix, D::Diagonal)
     Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
     ctranspose!(Ac, A)
     A_mul_B!(Ac, D)
 end
 
+At_mul_B(D::Diagonal, B::Diagonal) = Diagonal(D.diag .* B.diag)
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: IIRC `transpose` is recursive presently, requiring something along the lines of `Diagonal(transpose.(D.diag) .* B.diag)`?

File: base/linalg/diagonal.jl [hunk 3]
```diff
@@ -185,20 +185,23 @@ function A_mul_B!(D::Diagonal, B::UnitUpperTriangular)
...
 At_mul_B(A::AbstractTriangular, D::Diagonal) = A_mul_B!(transpose(A), D)
 function At_mul_B(A::AbstractMatrix, D::Diagonal)
     At = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
     transpose!(At, A)
     A_mul_B!(At, D)
 end
 
+A_mul_Bc(D::Diagonal, B::Diagonal) = Diagonal(D.diag .* conj(B.diag))
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: Likewise here, perhaps `conj.(...)` should replace `conj(...)`?

File: base/linalg/diagonal.jl [hunk 4]
```diff
@@ -207,13 +210,17 @@ function A_mul_Bc(D::Diagonal, A::AbstractMatrix)
     A_mul_B!(D, Ac)
 end
 
+A_mul_Bt(D::Diagonal, B::Diagonal) = At_mul_B(D, B)
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: I am not certain this definition is correct for non-scalar elements?

File: base/linalg/diagonal.jl [hunk 5]
```diff
@@ -207,13 +210,17 @@ function A_mul_Bc(D::Diagonal, A::AbstractMatrix)
...
 A_mul_Bt(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, transpose(B))
 function A_mul_Bt(D::Diagonal, A::AbstractMatrix)
     At = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
     transpose!(At, A)
     A_mul_B!(D, At)
 end
 
+Ac_mul_Bc(D::Diagonal, B::Diagonal) = Diagonal(conj(D.diag) .* conj(B.diag))
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: Likewise here, perhaps `conj.(...)` should replace `conj(...)`?

File: base/linalg/diagonal.jl [hunk 6]
```diff
@@ -207,13 +210,17 @@ function A_mul_Bc(D::Diagonal, A::AbstractMatrix)
...
+At_mul_Bt(D::Diagonal, B::Diagonal) = At_mul_B(D, B)
```
Comment (Sacha0) [inline base/linalg/diagonal.jl]: Please see above re. correctness for non-scalar elements? :)

File: base/linalg/diagonal.jl (hunk 2 above)
Comment (YingboMa) [inline base/linalg/diagonal.jl]: `At_mul_B(D::Diagonal, B::Diagonal) = Diagonal(transpose.(D.diag) .* B.diag)`? Sorry, I don't quite understand this.

Comment (Sacha0) [inline base/linalg/diagonal.jl]: Yes, and no worries! :) An illustration of the failure case:
```julia
julia> begin
           A = Diagonal([rand(2, 2), rand(2, 2)])
           B = Diagonal([rand(2, 2), rand(2, 2)])
           fullA = copy!(Matrix{Matrix{Float64}}(2, 2), A)
           fullB = copy!(Matrix{Matrix{Float64}}(2, 2), B)
       end;

julia> At_mul_B(fullA, fullB) == Diagonal(A.diag .* B.diag)
false

julia> At_mul_B(fullA, fullB) == Diagonal(transpose.(A.diag) .* B.diag)
true
```
Best!

Comment (YingboMa) [inline base/linalg/diagonal.jl]: Thanks for explaining!

Comment (YingboMa) [inline base/linalg/diagonal.jl]: Does `Ac_mul_B` better be `Ac_mul_B(D::Diagonal, B::Diagonal) = Diagonal(ctranspose.(D.diag) .* B.diag)` then?

Comment (Sacha0) [inline base/linalg/diagonal.jl]: Absolutely, and excellent catch! :)

File: test/linalg/diagonal.jl [hunk 7]
```diff
@@ -357,3 +357,19 @@ end
     @test logm(D) == Diagonal([logm([1 2; 3 4]), logm([1 2; 3 4])])
     @test sqrtm(D) == Diagonal([sqrtm([1 2; 3 4]), sqrtm([1 2; 3 4])])
 end
+
+@testset "multiplication of transposes of Diagonal (#22428)" begin
+    for t in (Float64, Complex{Float64})
+        D = Diagonal(randn(t, 5, 5))
+        B = Diagonal(randn(t, 5, 5))
+        @test D*B == Matrix(D)*Matrix(B) && typeof(D*B) <: Diagonal
```
Comment (fredrikekre) [inline test/linalg/diagonal.jl]: Not that it really matters but perhaps this is cleaner with a type assertion like:
```julia
@test (D*B)::Diagonal{t} == Matrix(D)*Matrix(B)
@test (D'B)::Diagonal{t} == Matrix(D)'*Matrix(B)
...
```

Comment (fredrikekre) [inline test/linalg/diagonal.jl]: Or even this:
```julia
for f in (*, Ac_mul_B, A_mul_Bc, Ac_mul_Bc, At_mul_B, A_mul_Bt, At_mul_Bt, Ac_mul_Bt, At_mul_Bc)
    @test f(D, B)::Diagonal{t} == f(Matrix(D), Matrix(B)) 
end
```

**Review by Sacha0 — APPROVED** [review]
The revisions look great! :) Perhaps add tests for the non-scalar-element case? (Edit: Sorry, this should have been a request for changes.)

**Review by Sacha0 — APPROVED** [review]
lgtm from rapid skim! :)

**Review by fredrikekre — APPROVED** [review]
Double timeout on Travis after relevant test passed. Thanks @YingboMa :)

**Pull request merged**